From 0b862efdd3965e2b82f7b65f437d87dafa84cae4 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Wed, 3 Oct 2018 02:45:16 -0400
Subject: [PATCH] OS-7256 smartos installer slow at drawing horizontal rule
 Reviewed by: Patrick Mooney <patrick.mooney@joyent.com> Reviewed by: Jorge
 Schrauwen <sjorge@blackdot.be>

---
 .../smartdc/lib/smartos_prompt_config.sh      | 53 ++++++++++++++++---
 1 file changed, 45 insertions(+), 8 deletions(-)

diff --git a/overlay/generic/smartdc/lib/smartos_prompt_config.sh b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
index 6a0b37ac..c682ab50 100644
--- a/overlay/generic/smartdc/lib/smartos_prompt_config.sh
+++ b/overlay/generic/smartdc/lib/smartos_prompt_config.sh
@@ -49,6 +49,36 @@ declare -a nics
 declare -a assigned
 declare prmpt_str
 
+#
+# Generate a horizontal ruler of appropriate length.  Doing this each time we
+# emit a ruler appears to induce a surprisingly noticeable latency, so we just
+# do it once.
+#
+ruler=
+while (( ${#ruler} < 80 )); do
+	ruler+='-'
+done
+
+#
+# Determine whether we need a trailing newline after the horizontal
+# ruler.  It
+# would seem that most modern terminal emulators behave like the VT100:
+# if the
+# cursor has naturally progressed to the rightmost column _and_ the next
+# character is a newline, the terminal will discard what is effectively
+# a
+# redundant newline.  Unfortunately, the framebuffer terminal emulator
+# in
+# illumos does not currently do this.  Attempt to guess if we are
+# attached to a
+# serial line or not, so as to decide whether we should send the extra
+# newline.
+#
+console=$(bootparams | awk -F= '$1 == "console" { print $2 }')
+if [[ $(/bin/tty) != '/dev/console' ]] || [[ $console =~ ^tty ]]; then
+	ruler+='\n'
+fi
+
 nicsup_done=0
 
 fatal()
@@ -576,6 +606,16 @@ updatenicstates()
 	done < <(dladm show-phys -po link,state 2>/dev/null)
 }
 
+printruler()
+{
+	#
+	# Print the horizontal ruler we have generated for this terminal.  Note
+	# that the ruler string is a printf(1) format string which may include
+	# a newline escape sequence.
+	#
+	printf -- "$ruler"
+}
+
 printheader()
 {
 	local newline=
@@ -583,25 +623,22 @@ printheader()
 	local subheader=$1
 
 	if [[ $(getanswer "simple_headers") == "true" ]]; then
-	echo "> ${subheader}"
-	return
-	fi
-
-	if [ $cols -gt 80 ] ;then
-	newline='\n'
+		echo "> ${subheader}"
+		return
 	fi
 
 	clear
 	printf " %-40s\n" "SmartOS Setup"
 	printf " %-40s%38s\n" "$subheader" "https://wiki.smartos.org/install"
-	for i in {1..80} ; do printf "-" ; done && printf "$newline"
+
+	printruler
 }
 
 print_warning()
 {
 	clear
 	printf "WARNING\n"
-	for i in {1..80} ; do printf "-" ; done && printf "\n"
+	printruler
 	printf "\n$1\n"
 
 	prmpt_str="\nPress [enter] to continue "
-- 
2.21.0

